presto-main-base/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java

```diff
@@ -395,6 +433,430 @@ public UpdateNode plan(Update node)
+                projectNodeWithUniqueRowIdsAndPresentColumn,
+                mergeAnalysis.getTargetTableScope(),
+                relationPlanWithUniqueRowIds.getFieldMappings());
+
+        RelationPlan sourceRelationPlan = new RelationPlanner(analysis, variableAllocator, idAllocator, lambdaDeclarationToVariableMap, metadata, session, sqlParser)
+                .process(mergeStmt.getSource(), sqlPlannerContext);
+
+        RelationPlan joinRelationPlan = new RelationPlanner(analysis, variableAllocator, idAllocator, lambdaDeclarationToVariableMap, metadata, session, sqlParser)
+                .planJoin(
+                        coerceIfNecessary(analysis, mergeStmt.getPredicate(), mergeStmt.getPredicate()),
+                        Join.Type.RIGHT, mergeAnalysis.getJoinScope(),
+                        planWithWithUniqueRowIdsAndPresentColumn, sourceRelationPlan,
+                        mergeStmt, sqlPlannerContext);
+
+        // Build the SearchedCaseExpression that creates the project "merge_row"
+        PlanBuilder joinSubPlan = newPlanBuilder(joinRelationPlan, analysis, lambdaDeclarationToVariableMap);
+
+        //    CASE
+        //       WHEN present THEN row(column1, column2, ..., present=true, operation UPDATE=3, case_number=0)
+        //       WHEN (present IS NULL) THEN row(column1, column2, ..., present=false, operation INSERT=1, case_number=1)
+        //       ELSE row(null, null, ..., false, -1, -1)
+        //    END
+        ImmutableList.Builder<WhenClause> whenClauses = ImmutableList.builder();
+        for (int caseNumber = 0; caseNumber < mergeStmt.getMergeCases().size(); caseNumber++) {
+            MergeCase mergeCase = mergeStmt.getMergeCases().get(caseNumber);
+
+            ImmutableList.Builder<Expression> joinResultBuilder = ImmutableList.builder();
+            List<List<ColumnHandle>> mergeCaseColumnsHandles = mergeAnalysis.getMergeCaseColumnHandles();
+            List<ColumnHandle> mergeCaseSetColumns = mergeCaseColumnsHandles.get(caseNumber);
+            for (ColumnHandle targetColumnHandle : mergeAnalysis.getTargetColumnHandles()) {
+                int index = mergeCaseSetColumns.indexOf(targetColumnHandle);
+                int fieldNumber = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(targetColumnHandle), "Field number for ColumnHandle is null");
+                Expression expression;
+                if (index >= 0) { // Update column value
+                    Expression setExpression = mergeCase.getSetExpressions().get(index);
+                    joinSubPlan = subqueryPlanner.handleSubqueries(joinSubPlan, setExpression, mergeStmt, sqlPlannerContext);
+                    expression = joinSubPlan.rewrite(setExpression);
+                    expression = coerceIfNecessary(analysis, setExpression, expression);
+                    expression = checkNotNullColumns(targetColumnHandle, expression, fieldNumber, mergeAnalysis);
```

Comment:
I don't think this should be done at the plan level, as it will be very expensive.  Instead, we should send to the MergeWriterOperator what columns are non-null, and the operator can do a more efficient bulk check.  Let me know if I misunderstood anything.